File: polynomial_secret.js
const fs = require('fs');

// Function to decode a value from given base to decimal
function decodeValue(value, base) {
    // For bases > 10, we need to handle letters
    if (base <= 10) {
        return BigInt(parseInt(value, base));
    } else {
        // Convert using BigInt for large numbers
        let result = 0n;
        const baseBI = BigInt(base);
        
        for (let i = 0; i < value.length; i++) {
            const char = value[i];
            let digit;
            
            if (char >= '0' && char <= '9') {
                digit = BigInt(char.charCodeAt(0) - '0'.charCodeAt(0));
            } else if (char >= 'a' && char <= 'z') {
                digit = BigInt(10 + char.charCodeAt(0) - 'a'.charCodeAt(0));
            } else if (char >= 'A' && char <= 'Z') {
                digit = BigInt(10 + char.charCodeAt(0) - 'A'.charCodeAt(0));
            }
            
            result = result * baseBI + digit;
        }
        
        return result;
    }
}

// Function to parse test case and extract points
function extractPoints(testCase) {
    const points = [];
    
    for (const [key, value] of Object.entries(testCase)) {
        if (key === 'keys') continue;
        
        const x = BigInt(key);
        const base = parseInt(value.base);
        const y = decodeValue(value.value, base);
        
        points.push({ x, y });
    }
    
    return points;
}

// Extended Euclidean Algorithm for modular inverse
function modInverse(a, m) {
    const [g, x] = extendedGCD(a, m);
    if (g !== 1n) throw new Error('Modular inverse does not exist');
    return ((x % m) + m) % m;
}

function extendedGCD(a, b) {
    if (a === 0n) return [b, 0n, 1n];
    const [g, x1, y1] = extendedGCD(b % a, a);
    const x = y1 - (b / a) * x1;
    const y = x1;
    return [g, x, y];
}

// GCD function for fraction reduction
function gcd(a, b) {
    a = a < 0n ? -a : a;
    b = b < 0n ? -b : b;
    while (b !== 0n) {
        const temp = b;
        b = a % b;
        a = temp;
    }
    return a;
}

// Lagrange interpolation to find f(0) (the constant term)
function lagrangeInterpolation(points, k) {
    // Use only the first k points
    const selectedPoints = points.slice(0, k);
    
    // We'll accumulate the result as a fraction (numerator/denominator)
    let totalNumerator = 0n;
    let totalDenominator = 1n;
    
    for (let i = 0; i < selectedPoints.length; i++) {
        // Calculate Li(0) = product of (-xj)/(xi - xj) for all j != i
        let termNumerator = selectedPoints[i].y;
        let termDenominator = 1n;
        
        for (let j = 0; j < selectedPoints.length; j++) {
            if (i !== j) {
                // numerator *= -xj
                termNumerator *= -selectedPoints[j].x;
                // denominator *= (xi - xj)
                termDenominator *= (selectedPoints[i].x - selectedPoints[j].x);
            }
        }
        
        // Add this term to our total: total = total + term
        // (a/b) + (c/d) = (a*d + b*c) / (b*d)
        totalNumerator = totalNumerator * termDenominator + totalDenominator * termNumerator;
        totalDenominator = totalDenominator * termDenominator;
        
        // Reduce the fraction to prevent overflow
        const g = gcd(totalNumerator, totalDenominator);
        totalNumerator = totalNumerator / g;
        totalDenominator = totalDenominator / g;
    }
    
    // The result should be an integer for this problem
    if (totalDenominator !== 1n && totalNumerator % totalDenominator !== 0n) {
        throw new Error('Result is not an integer');
    }
    
    return totalNumerator / totalDenominator;
}

// Main function
function main() {
    const args = process.argv.slice(2);
    if (args.length < 1) {
        console.log('Usage: node polynomial_secret.js <testcase.json>');
        return;
    }
    
    const filename = args[0];
    
    try {
        // Read and parse the test case
        const data = fs.readFileSync(filename, 'utf8');
        const testCase = JSON.parse(data);
        
        console.log(`Test Case: ${filename}`);
        console.log(`Number of roots (n): ${testCase.keys.n}`);
        console.log(`Minimum roots required (k): ${testCase.keys.k}`);
        console.log('\nDecoded Points:');
        
        // Extract and decode points
        const points = extractPoints(testCase);
        
        // Print decoded points
        points.forEach(point => {
            console.log(`  x = ${point.x}, y = ${point.y}`);
        });
        
        // Find the secret using Lagrange interpolation
        const secret = lagrangeInterpolation(points, testCase.keys.k);
        console.log(`\nSecret (constant term c): ${secret}`);
        
    } catch (error) {
        console.error(`Error: ${error.message}`);
    }
}

// Run the main function
main();

File: polynomial_secret.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"math/big"
	"os"
	"strconv"
)

// TestCase represents the JSON structure
type TestCase struct {
	Keys struct {
		N int `json:"n"`
		K int `json:"k"`
	} `json:"keys"`
	Roots map[string]struct {
		Base  string `json:"base"`
		Value string `json:"value"`
	} `json:"-"`
}

// Point represents a decoded (x, y) point
type Point struct {
	X *big.Int
	Y *big.Int
}

// UnmarshalJSON custom unmarshaler to handle dynamic root keys
func (t *TestCase) UnmarshalJSON(data []byte) error {
	// First unmarshal into a map
	var raw map[string]json.RawMessage
	if err := json.Unmarshal(data, &raw); err != nil {
		return err
	}

	// Parse keys
	if keysData, ok := raw["keys"]; ok {
		if err := json.Unmarshal(keysData, &t.Keys); err != nil {
			return err
		}
	}

	// Parse roots
	t.Roots = make(map[string]struct {
		Base  string `json:"base"`
		Value string `json:"value"`
	})

	for key, value := range raw {
		if key == "keys" {
			continue
		}
		var root struct {
			Base  string `json:"base"`
			Value string `json:"value"`
		}
		if err := json.Unmarshal(value, &root); err == nil {
			t.Roots[key] = root
		}
	}

	return nil
}

// decodeValue decodes a value from given base to decimal
func decodeValue(value string, base int) (*big.Int, error) {
	result := new(big.Int)
	result.SetString(value, base)
	return result, nil
}

// parseTestCase reads and parses the JSON test case
func parseTestCase(filename string) (*TestCase, error) {
	data, err := ioutil.ReadFile(filename)
	if err != nil {
		return nil, err
	}

	var testCase TestCase
	if err := json.Unmarshal(data, &testCase); err != nil {
		return nil, err
	}

	return &testCase, nil
}

// extractPoints converts the test case into decoded points
func extractPoints(testCase *TestCase) ([]Point, error) {
	points := make([]Point, 0)

	for xStr, root := range testCase.Roots {
		// Parse X value
		x := new(big.Int)
		x.SetString(xStr, 10)

		// Parse base
		base, err := strconv.Atoi(root.Base)
		if err != nil {
			return nil, fmt.Errorf("invalid base: %s", root.Base)
		}

		// Decode Y value
		y, err := decodeValue(root.Value, base)
		if err != nil {
			return nil, fmt.Errorf("failed to decode value: %s in base %d", root.Value, base)
		}

		points = append(points, Point{X: x, Y: y})
	}

	return points, nil
}

// lagrangeInterpolation performs Lagrange interpolation to find f(0)
func lagrangeInterpolation(points []Point, k int) *big.Int {
	// We only need the first k points
	if len(points) > k {
		points = points[:k]
	}

	result := new(big.Int)

	for i := 0; i < len(points); i++ {
		// Calculate Li(0) = product of (-xj)/(xi - xj) for all j != i
		numerator := new(big.Int).Set(points[i].Y)
		denominator := big.NewInt(1)

		for j := 0; j < len(points); j++ {
			if i != j {
				// numerator *= -xj
				negXj := new(big.Int).Neg(points[j].X)
				numerator.Mul(numerator, negXj)

				// denominator *= (xi - xj)
				diff := new(big.Int).Sub(points[i].X, points[j].X)
				denominator.Mul(denominator, diff)
			}
		}

		// Add the contribution to result
		term := new(big.Int).Div(numerator, denominator)
		result.Add(result, term)
	}

	return result
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Usage: go run polynomial_secret.go <testcase.json>")
		return
	}

	filename := os.Args[1]

	// Parse test case
	testCase, err := parseTestCase(filename)
	if err != nil {
		fmt.Printf("Error parsing test case: %v\n", err)
		return
	}

	fmt.Printf("Test Case: %s\n", filename)
	fmt.Printf("Number of roots (n): %d\n", testCase.Keys.N)
	fmt.Printf("Minimum roots required (k): %d\n", testCase.Keys.K)
	fmt.Println("\nDecoded Points:")

	// Extract and decode points
	points, err := extractPoints(testCase)
	if err != nil {
		fmt.Printf("Error extracting points: %v\n", err)
		return
	}

	// Print decoded points
	for _, point := range points {
		fmt.Printf("  x = %s, y = %s\n", point.X.String(), point.Y.String())
	}

	// Find the secret using Lagrange interpolation
	secret := lagrangeInterpolation(points, testCase.Keys.K)
	fmt.Printf("\nSecret (constant term c): %s\n", secret.String())
}
